pkg/distributor/query.go

```diff
@@ -138,38 +138,59 @@ func (d *Distributor) queryIngesterStream(ctx context.Context, replicationSet ri
 		}
 		defer stream.CloseSend() //nolint:errcheck
 
-		var result []*ingester_client.QueryStreamResponse
+		result := &ingester_client.QueryStreamResponse{}
 		for {
-			series, err := stream.Recv()
+			resp, err := stream.Recv()
 			if err == io.EOF {
 				break
 			} else if err != nil {
 				return nil, err
 			}
-			result = append(result, series)
+
+			result.Chunkseries = append(result.Chunkseries, resp.Chunkseries...)
+			result.Timeseries = append(result.Timeseries, resp.Timeseries...)
 		}
 		return result, nil
 	})
 	if err != nil {
 		return nil, err
 	}
 
-	hashToSeries := map[model.Fingerprint]ingester_client.TimeSeriesChunk{}
+	hashToChunkseries := map[model.Fingerprint]ingester_client.TimeSeriesChunk{}
+	hashToTimeSeries := map[model.Fingerprint]ingester_client.TimeSeries{}
+
 	for _, result := range results {
-		for _, response := range result.([]*ingester_client.QueryStreamResponse) {
-			for _, series := range response.Timeseries {
-				hash := client.FastFingerprint(series.Labels)
-				existing := hashToSeries[hash]
-				existing.Labels = series.Labels
-				existing.Chunks = append(existing.Chunks, series.Chunks...)
-				hashToSeries[hash] = existing
-			}
+		response := result.(*ingester_client.QueryStreamResponse)
+
+		// Parse any chunk series
+		for _, series := range response.Chunkseries {
+			hash := client.FastFingerprint(series.Labels)
+			existing := hashToChunkseries[hash]
+			existing.Labels = series.Labels
+			existing.Chunks = append(existing.Chunks, series.Chunks...)
+			hashToChunkseries[hash] = existing
+		}
+
+		// Parse any time series
+		for _, series := range response.Timeseries {
+			hash := client.FastFingerprint(series.Labels)
+			existing := hashToTimeSeries[hash]
+			existing.Labels = series.Labels
+			existing.Samples = append(existing.Samples, series.Samples...)
```

Comment:
Isn't samples ordering required here? Because then series these goes into `newTimeSeriesSeriesSet()`, whose iterator `Seek()` requires samples sorted, but I can't find where such sorting occurs.

/cc @pstibrany

Reply:
Good catch Marco!

My assumption was always that entire `SeriesIterator` (and now `chunkenc.Iterator`) expects samples to be sorted, but interestingly, it doesn't say so in the documentation. Hmm.

Reply:
Created https://github.com/prometheus/prometheus/issues/6876 to get clarification.

Reply:
Yes, it must be sorted.

Reply:
Thanks Brian!

Reply:
So we need to sort these before returning? Wouldn't the chunks need to have been sorted before my changes as well?

Reply:
I've added a sort before returning the timeseries

Reply:
> So we need to sort these before returning? Wouldn't the chunks need to have been sorted before my changes as well?

This is a great question! In case of chunks, we have three different "chunk merging" functions (see `getChunksIteratorFunction` function in querier.go). Two of them (`batch.NewChunkMergeIterator` and `iterators.NewChunkMergeIterator`) sort chunks by From time before using them, while third one (`mergeChunks`) simply gets all the samples from chunks, and then sorts samples).

So extra sorting is not needed for chunks here, and is done later when merging chunks together.